### PR TITLE
[MRV2][Spec Decode] DFlash for MRV2

### PR DESCRIPTION
**Pull request by benchislett**
## Purpose

This PR implements DFlash for ModelRunnerV2. We are still working to refine the design, and notably this PR replicates a lot of code with the existing EAGLE codepath. Points of interest have been highlighted as PR comments.

## Testing

I ran the existing acceptance rate test for DFlash with MRV1 vs MRV2 and saw no differences. Still need to manually validate that the gumbel probabilistic sampling is working properly.

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/gumbel.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ def gumbel_block_argmax(
     vocab_size,
     APPLY_TEMPERATURE: tl.constexpr,
     USE_FP64: tl.constexpr,
+    PER_TOKEN_COL: tl.constexpr = False,
```
Comment (benchislett) [inline vllm/v1/worker/gpu/sample/gumbel.py:94]: This is a change I added for the gumbel sampler. I don't fully understand how the sampler works, @WoosukKwon please review if this can be done more elegantly or if it's maybe not needed in the first place.

The main idea is that DFlash needs to sample in parallel as opposed to EAGLE's autoregressive drafting, and I want to make sure that we use the same seed per-position so that probabilistic acceptance still works.

File: vllm/v1/worker/gpu/spec_decode/dflash/cudagraph.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            query_start_loc_gpu=input_batch.query_start_loc,
+            query_start_loc_cpu=query_start_loc_cpu,
+            max_query_len=num_tokens // num_reqs,
+            seq_lens=input_batch.seq_lens,
+            max_seq_len=max_model_len,
+            block_tables=input_block_tables,
+            slot_mappings=slot_mappings,
+            kv_cache_config=kv_cache_config,
+            for_cudagraph_capture=True,
+            causal=False,
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/cudagraph.py:57]: Main difference here is just setting causal=False, this helper could probably be shared with EAGLE

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+        # We need to get the hidden size from the draft model config because
+        # the draft model's hidden size can be different from the target model's
+        # hidden size (e.g., Llama 3.3 70B).
+        self.hidden_size = self.draft_model_config.get_hidden_size()
+        self.vocab_size = self.draft_model_config.get_vocab_size()
+        self.dtype = vllm_config.model_config.dtype
+        self.use_fp64_gumbel = vllm_config.model_config.use_fp64_gumbel
+
+        # DP configuration
+        self.dp_size = vllm_config.parallel_config.data_parallel_size
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:56]: A lot of the DP/CG/Sampling state here could be shared with EAGLE in a common base class

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+        self.query_cudagraph_manager: DFlashCudaGraphManager | None = None
+        self.draft_kv_cache_group_id: int = -1
+
+        # Multimodal inputs not currently supported.
+        self.supports_mm_inputs = False
+
+    def init_cudagraph_manager(self, cudagraph_mode: CUDAGraphMode) -> None:
+        cudagraph_mode = self.vllm_config.compilation_config.cudagraph_mode
+
+        # PIECEWISE cudagraphs are not supported for dflash draft forwards.
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:150]: @TheEpicDolphin I'm hoping that this constraint could be relieved but haven't checked it out

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+            decode_query_len=self.num_query_per_req,
+        )
+
+    def load_model(self, target_model: nn.Module) -> None:
+        target_attn_layer_names = get_layers_from_vllm_config(
+            self.vllm_config,
+            AttentionLayerBase,  # type: ignore[type-abstract]
+        ).keys()
+
+        self.model = load_dflash_model(target_model, self.vllm_config)
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:172]: This is basically the same as the EAGLE variant

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+            kv_cache_config,
+            self.vllm_config,
+            self.device,
+            active_layer_names=self.draft_attn_layer_names,
+        )
+        self.block_tables = block_tables
+
+        # DFlash precomputes context K/V with a single block_size; mixing
+        # kv-cache groups would silently corrupt the cache for the non-matching group.
+        draft_groups = [gid for gid, g in enumerate(self.attn_groups) if g]
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:200]: Extra assertion here because we have a fused KV cache preprocessing step for DFlash. Not sure which if any models require multiple draft KV groups

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+                : num_reqs_padded + 1
+            ],
+            query_start_loc_cpu=query_start_loc_cpu,
+            max_query_len=self.num_query_per_req,
+            seq_lens=self.input_buffers.seq_lens[:num_reqs_padded],
+            max_seq_len=self.max_model_len,
+            block_tables=block_tables,
+            slot_mappings=slot_mappings,
+            kv_cache_config=self.kv_cache_config,
+            causal=False,
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:319]: More duplication here, causal=False is the main delta

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+        dummy_run: bool = False,
+        skip_attn_for_dummy_run: bool = False,
+        mm_inputs: tuple[list[torch.Tensor], torch.Tensor] | None = None,
+        is_profile: bool = False,
+    ) -> torch.Tensor:
+        num_reqs = input_batch.num_reqs
+        num_target_tokens = input_batch.num_tokens
+        num_query_tokens = num_reqs * self.num_query_per_req
+
+        # NOTE(woosuk): To avoid CPU-GPU synchronization without CPU knowing the
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:371]: Lots of shared code in propose(). I copied the comments too in order to minimize the diff.

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+    q_slot = q_block_id * block_size + (query_pos % block_size)
+
+    tl.store(out_input_ids_ptr + query_idx, input_id, mask=is_query)
+    tl.store(out_query_positions_ptr + query_idx, query_pos, mask=is_query)
+    tl.store(out_query_slot_mapping_ptr + query_idx, q_slot, mask=is_query)
+
+    # --- Sample indices / positions / idx_mapping (mask tokens only) ---
+    is_sample = is_query & (query_off > 0)
+    sample_idx = req_idx * num_speculative_steps + (query_off - 1)
+    tl.store(out_sample_indices_ptr + sample_idx, query_idx, mask=is_sample)
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:589]: TODO: I think this can just be a constant precomputed arange `[[1, 2, 3, ..., n-1], [1, 2, 3, ..., n-1], ...]` or just a runtime slice
